test/unit/contrib/parallel/reader_writer_latch_test.cpp

```diff
@@ -0,0 +1,98 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+
+#include <seqan3/contrib/parallel/buffer_queue.hpp>
+#include <seqan3/contrib/parallel/buffer_queue.hpp>
+#include <seqan3/contrib/parallel/reader_writer_latch.hpp>
+
+using namespace seqan3::contrib;
+
+TEST(reader_writer_latch, parallel)
+{
+    size_t threads = std::thread::hardware_concurrency();
+
+    if (threads > 4)
+        threads = 4;
+
+    if (threads > 1)
+        --threads;  // One less for the parallel reader
+
+    EXPECT_GE(threads, 1u);
+
+    size_t job_size = threads * 1000000;
+
+    fixed_buffer_queue<uint32_t> source_queue{job_size};
+    fixed_buffer_queue<uint32_t> target_queue{job_size};
+
+    reader_writer_latch source_latch{reader_count{threads}, writer_count{1}, source_queue};
+    reader_writer_latch target_latch{reader_count{1}, writer_count{threads}, target_queue};
+
+    auto work = [&] () mutable
+    {
+        {
+            scoped_reader_guard reader_guard{source_latch};
+            scoped_writer_guard writer_guard{target_latch};
+
+            for (;;)
+            {
+                uint32_t val = 0;
+                if (source_queue.wait_pop(val) == queue_op_status::closed)
+                    return;
+
+                queue_op_status status = target_queue.try_push(val);
+                EXPECT_TRUE(status == queue_op_status::success);
+            }
+        }
+    };
+
+    uint32_t counter{0};
+
+    // start the producer of source/consumer of target.
+    std::thread t1{[&] ()
+    {
+        {
+            scoped_writer_guard writer_guard{source_latch};
+
+            // Initialise source_queue.
+            for (size_t i = 0; i < job_size; ++i)
+            {
+                queue_op_status status = source_queue.try_push(1);
```

Comment:
I would here sum up `i`, because the same value might hide side-effects within tested functions (for example reads the same value concurrently; even if already consumed).